### PR TITLE
Allow unset `--entrypoint` in `docker run` or `docker create`

### LINKED ISSUE #23498: Allow `--entrypoint` to be set to `null` in a `docker run` or `docker create`
I would like to be able to "unset" an ENTRYPOINT at the command line when creating a container.

Using this Dockerfile and entrypoint.sh as an example:

Dockerfile:

    FROM alpine
    ADD entrypoint.sh /entrypoint.sh
    RUN chmod 755 /entrypoint.sh
    ENTRYPOINT ["/entrypoint.sh"]
    CMD echo foobar

entrypoint:

    #!/bin/sh
    echo "I am an entrypoint"
    exec "$@"

If I wanted to unset the entrypoint, I would like to be able to run this: `docker run --entrypoint='' --rm -it image echo foo`

I would expect this to produce the following output:

    foo

but it does the following instead:

    I am an entrypoint
    foo

Providing some way for `docker run` or `docker create` to force the ENTRYPOINT to be null without having to create another image would be great.

### PR DESCRIPTION
**Pull request by yongtang**
**\- What I did**

This fix tries to address the issue raised in #23498 to allow unset `--entrypoint` in `docker run` or `docker create`.

**\- How I did it**

This fix checks the flag `--entrypoint` and, in case `--entrypoint=` (`""`) is passed, unset the Entrypoint during the container run.

**\- How to verify it**

Additional integration tests have been created to cover changes in this fix.

**\- Description for the changelog**

Allow unset `--entrypoint` in `docker run` or `docker create` by passing `--entrypoint=`.

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #23498.

Signed-off-by: Yong Tang yong.tang.github@outlook.com

### CODE REVIEW

**Comment by thaJeztah** [comment]
Yes, looks good

**Comment by LK4D4** [comment]
@yongtang would you mind to rebase?

**Comment by yongtang** [comment]
@LK4D4 Sure let me rebase it shortly. 😃

**Comment by yongtang** [comment]
@LK4D4 Rebased. Thanks.

**Comment by vdemeester** [comment]
LGTM 🐸
This is gonna need updates in docs too 👼

**Comment by tonistiigi** [comment]
LGTM

**Comment by yongtang** [comment]
@tonistiigi @vdemeester @LK4D4 @thaJeztah The pull request has been updated with docs in `docs/reference/api/docker_remote_api_v1.25.md` and `docs/reference/run.md`. Please take a look and let me know if there are any issues.

File: docs/reference/run.md [hunk 1]
```diff
@@ -1308,6 +1308,10 @@ or two examples of how to pass more parameters to that ENTRYPOINT:
 > **Note**: Passing `--entrypoint` will clear out any default command set on the
 > image (i.e. any `CMD` instruction in the Dockerfile used to build it).
 
+> **Note**: Passing `--entrypoint=""` (empty string) will reset the entry
+> point to system default (i.e., the entry point used by docker when there
+> is no `ENTRYPOINT` instruction in the Dockerfile).
```
Comment (thaJeztah) [inline docs/reference/run.md]: I don't think this should be a note (as it's not an "unexpected" behavior) perhaps similar to the flag description;

```Markdown
You can reset a containers entrypoint by passing an empty string, for example:

     $ docker run -it --entrypoint="" mysql bash

```

(and move that above the `**Note**`)

Comment (thaJeztah) [inline docs/reference/run.md]: Wondering; I'm not at my computer now, but does resetting the entrypoint also reset the `CMD` (the same as setting a custom entrypoint?)

If so, I think we should have a test case for that

Comment (yongtang) [inline docs/reference/run.md]: Thanks @thaJeztah. The pull request has been updated with the documentation change (like a flag description).

Resetting the entrypoint will also reset `CMD`. I adds an additional test case for that. Please let me know if there are any issues.

**Comment by vdemeester** [comment]
LGTM 🐸

**Comment by thaJeztah** [comment]
LGTM, thanks!

**Pull request merged**